src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -99,6 +108,54 @@ UniqueTypeName XdsClusterAttribute::TypeName() {
   return kFactory.Create();
 }
 
+namespace {
+class XdsResolver;
+}  // namespace
+
+// An entry in the map of clusters that need to be present in the LB
+// policy config.  The map holds a weak ref.  One strong ref is held by
+// the ConfigSelector, and another is held by each call assigned to
+// the cluster by the ConfigSelector.  The ref for each call is held
+// until the call is committed.  When the strong refs go away, we hop
+// back into the WorkSerializer to remove the entry from the map.
+class ClusterState : public DualRefCounted<ClusterState> {
+ public:
+  using ClusterStateMap =
+      std::map<std::string, WeakRefCountedPtr<ClusterState>>;
+
+  ClusterState(RefCountedPtr<XdsResolver> resolver,
+               const std::string& cluster_name);
+
+  void Orphan() override;
+
+  const std::string& cluster() const { return it_->first; }
+
+ private:
+  RefCountedPtr<XdsResolver> resolver_;
+  ClusterStateMap::iterator it_;
+};
+
+class XdsClusterMap : public RefCounted<XdsClusterMap> {
```

Comment:
Please add a comment here explaining what this class is for.  In particular, it's the cluster map being held by the `XdsConfigSelector`, not the one being held by the resolver.  Refs to this will also be held by individual calls between the time that the `XdsConfigSelector` is called and the time that the `ClusterSelectionFilter` runs.